content/blog/2022/2022-08-30-handling-timezones-in-spring.md

```diff
@@ -0,0 +1,516 @@
+
+{{% github "https://github.com/thombergs/code-examples/tree/master/spring-boot/timezones" %}}
+
+## Understanding GMT, UTC and DST
+- **Greenwich Mean Time (GMT)** is a timezone used in some parts of the world, mainly Europe and Africa.
+GMT was replaced as the international standard time in 1972 by UTC.
+- **Universal Coordinated Time (UTC)** is not a timezone. It is a universally preferred standard that can be used to display timezones.
+- **Daylight Savings Time (DST)** is the practice of setting clocks forward by one hour in the summer months
+and back again in the fall, to make better use of natural daylight. **Neither GMT or UTC get affected by DST**. To account for DST changes,
+countries or states usually switch to another timezone. For instance in Australian summer, the states that observe DST will move from
+Australian Eastern Standard Time (AEST) to Australian Eastern Daylight Time (AEDT).
+
+Operations around dates, time and timezones can be confusing and prone to errors. To understand some problems around dates refer to this [article.](https://yourcalendricalfallacyis.com/)
+We will deep-dive into various aspects of handling timezones in the further sections.
+
+## Drawbacks of legacy time-based java.util classes
+
+Let's look at a few reasons why you should choose to avoid the date-time classes in the java.util package when developing applications.
+
+### Missing Timezone information
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/date.JPG" %}}
+
+- `java.util.Date` represents an instant in time. 
+- Also, it has no timezone information. So, it considers a **default system timezone** which could differ from location to location.
+For instance, if a person runs this test from another country, he might see a different date and time derived from the given milliseconds.
+
+### Custom Date creation
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/customDate.JPG" %}}
+
+- Creating a custom date with this API is very inconvenient. Firstly, the year starts with 1900, hence we must
+subtract 1900 so that the right year is considered.
+- Also to derive the months we need to use indexes 0-11. In this example, to create a date in August we would use 7 and not 8.
+
+### Mutable Classes
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/mutableDate.JPG" %}}
+
+- Immutability is a key concept that ensures that java objects are thread-safe and concurrent access does not lead to an inconsistent state.
+- The Date API **allows mutable methods** such as `setHours`, `setMinutes`, `setDate`. Therefore, it becomes the **responsibility
+of the developer to clone the object before use**.
+- Similarly, the Calendar object also has setter methods `setTimeZone`, `add` which allows an object to be modified.
+
+### Format Dates
+
+{{% image alt="settings" src="images/posts/handling-timezones-in-spring/dateFormat.JPG" %}}
```

Comment:
Same here (and for all other screenshots that can be converted into code, I'm not going to point them all out).